### PR TITLE
Tensorrt development support

### LINKED ISSUE #92: [FEATURE] Provide docker for development with onnxruntime-tensorrt
**Is your feature request related to a problem? Please describe.**
Enable validation and reporting for onnxruntime-tensorrt

### LINKED ISSUE #93: [BUG] Misleading prediction time profiling for vanilla pytorch validation
**Describe the bug**
Prediction time profiling for vanilla pytorch validation doesn't wait for synchronizaiton

**To Reproduce**
1. Run validation pytorch model

**Expected behavior**
Prediction time profiling for vanilla pytorch validation should includes synchronizaiton

### PR DESCRIPTION
**Pull request by alifahrri**
## Type of changes

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Tensorrt dockerfile doesnt support development env. Misleading prediction time profile for pytorch model validation.

fix #93, close #92 

## What is the new behavior?

- Tensorrt dockerfile support development env.
- prediction time profile for pytorch model validation includes synchronization

## Checklist

- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/development/vortex/development/core/engine/validator/base_validator.py [hunk 1]
```diff
@@ -308,7 +308,7 @@ def torch_predict(cls, predictor: torch.nn.Module, image, batch_size=None, *args
             if isinstance(kwargs[i], np.ndarray) :
                 kwargs[i] = torch.from_numpy(kwargs[i])
             kwargs[i] = kwargs[i].to(device)
-        return predictor(image, *args, **kwargs)
+        return predictor(image, *args, **kwargs).cpu().numpy()
```
[the remark this answers is not in the document]
Comment (triwahyuu) [reply src/development/vortex/development/core/engine/validator/base_validator.py:311]: https://pytorch.org/docs/stable/cuda.html#torch.cuda.synchronize

File: dockerfiles/onnxruntime-tensorrt.dockerfile [hunk 2]
```diff
@@ -32,14 +48,19 @@ RUN git clone --single-branch --branch ${ONNXRUNTIME_BRANCH} --recursive ${ONNXR
...
 # need to manually install onnx instead of using extras
-RUN cd src/runtime && pip install onnx==1.6.0 && pip install -e .
\ No newline at end of file
+RUN cd src/runtime && pip install -U setuptools onnx==1.6.0 && pip install -e .
+
+FROM runtime AS development
+# install vortex-development
+COPY src/development src/development
+RUN pip install --upgrade pip
+RUN cd src/development && pip install --ignore-installed --timeout=10000 -e .
```
Comment (triwahyuu) [inline dockerfiles/onnxruntime-tensorrt.dockerfile:66]: could you please add testing to check that votex is installed and tensorrt runtime is available?

**Review by triwahyuu — COMMENTED** [review]
could you please add docker build ci?

File: dockerfiles/onnxruntime-tensorrt.dockerfile (hunk 2 above)
Comment (alifahrri) [reply dockerfiles/onnxruntime-tensorrt.dockerfile:66]: moving tensor to cpu actually performs synchronization https://pytorch.org/docs/stable/notes/cuda.html#asynchronous-execution

> PyTorch automatically performs necessary synchronization when copying data between CPU and GPU or between two GPUs

so i think it is better let pytorch decide the synchronization

File: src/development/vortex/development/core/engine/validator/base_validator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alifahrri) [reply src/development/vortex/development/core/engine/validator/base_validator.py:311]: afaik moving tensor to cpu actually performs synchronization https://pytorch.org/docs/stable/notes/cuda.html#asynchronous-execution

> PyTorch automatically performs necessary synchronization when copying data between CPU and GPU or between two GPUs

so i think it is better let pytorch decide the synchronization

**Comment by alifahrri** [comment]
> could you please add docker build ci?

ok brb fix

File: src/development/vortex/development/core/engine/validator/base_validator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (triwahyuu) [reply src/development/vortex/development/core/engine/validator/base_validator.py:311]: ohh I see, okayy then, I didn't know about that

File: dockerfiles/onnxruntime-tensorrt.dockerfile (hunk 2 above)
Comment (triwahyuu) [reply dockerfiles/onnxruntime-tensorrt.dockerfile:66]: okay

Comment (alifahrri) [reply dockerfiles/onnxruntime-tensorrt.dockerfile]: updated

**Comment by triwahyuu** [comment]
is it possible for the development tag to have all runtime including the tensorrt runtime? then the development build from the main dockerfile could be removed. I think that would be easier to use and not having too many tags each release

**Comment by alifahrri** [comment]
> is it possible for the development tag to have all runtime including the tensorrt runtime? then the development build from the main dockerfile could be removed. I think that would be easier to use and not having too many tags each release

the development targget from onnxruntime-tensorrt actually has all runtime (onnxruntime tensorrt, cuda, cpu, and torchscript) available, but additional feature (hyperparam opt) not included

**Comment by triwahyuu** [comment]
as I understand from the docker build tensorrt for the development, it will overwrite the main vortex development build since it only use the version as tag, and also for latest (e.g. [nodefluxio/vortex:0.2.1](https://hub.docker.com/layers/nodefluxio/vortex/0.2.1/images/sha256-16282a5fbb483fe74dd0761514a6074fc4540d8b2463c0fa156d45fdf234f773?context=explore)). won't it be better to just fully support it? since it will be actually be the main images.

**Comment by alifahrri** [comment]
may be i misunderstood tag in the workflow file, turns out it is just vortex version like you said. But anyway I think it is better to support separately at least for now since this kind of workflow took to long to build

**Comment by triwahyuu** [comment]
oh I see, okayy. then you will change the tag name for the tensorrt development right?

File: .github/workflows/docker_build_trt.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+          echo "RELEASE_VERSION=$(${{ github.event_name == 'release' }} && echo "${VER//v}" || echo "$VER")" >> $GITHUB_ENV
+
+      - uses: docker/setup-buildx-action@v1
+      - name: Build and Publish Docker Image
+        uses: docker/build-push-action@v1.1.0
+        with:
+          repository: nodefluxio/vortex
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+          tags: "${{ env.RELEASE_VERSION }},trt-latest"
```
Comment (alifahrri) [inline .github/workflows/docker_build_trt.yml:61]: @triwahyuu continuing from comment, for tagging prefer to make it separate from main, so I think this liine should be
```
"development-${{ matrix.runtime_type }}-${{ env.RELEASE_VERSION }},latest"
```
or smth like that ?

Comment (triwahyuu) [reply .github/workflows/docker_build_trt.yml:61]: maybe this one doesn't need `latest` tag, as it would be for the main vortex development image. 
my reasoning to not include `development` in the tag is that because the image actually contains both development and runtime so I think it's the same anyway.

maybe for the vortex development with tensorrt the tag is just `${{ env.RELEASE_VERSION }}-tensorrt`, what do you think?

Comment (alifahrri) [reply .github/workflows/docker_build_trt.yml:61]: ok that seems good brb change it

**Comment by triwahyuu** [comment]
it seems like you're not using paralel compilation when compiling the onnxruntime, that's why it takes way too long to complete.
I think you could use the [`--parallel`](https://github.com/microsoft/onnxruntime/blob/f874260b9e0d26a603b64e15da5a454a4b67c229/tools/ci_build/build.py#L142) argument for it. 
reference from this https://github.com/microsoft/onnxruntime/blob/master/build.sh

File: dockerfiles/onnxruntime-tensorrt.dockerfile [hunk 4]
```diff
@@ -7,21 +7,37 @@
...
+            wget \
+            zip \
+            ca-certificates \
+            build-essential \
+            curl \
+            libcurl4-openssl-dev \
+            libssl-dev \
+            python3-dev
+
+RUN wget --quiet https://repo.anaconda.com/miniconda/Miniconda3-4.5.4-Linux-x86_64.sh -O ~/miniconda.sh --no-check-certificate && /bin/bash ~/miniconda.sh -b -p /opt/miniconda && \
```
Comment (triwahyuu) [inline dockerfiles/onnxruntime-tensorrt.dockerfile:30]: I think it would be best to use the same installation process as in the [main docker file](https://github.com/nodefluxio/vortex/blob/652d04705acdf92768ad36f2a347de5f1e75bc3f/Dockerfile#L19) for consistency. and also it would be better to install pytorch in here using conda including the corresponding cudatoolkit version to avoid errors as you have suggested before.

File: dockerfiles/onnxruntime-tensorrt.dockerfile [hunk 5]
```diff
@@ -7,21 +7,37 @@
...
 ARG ONNXRUNTIME_BRANCH=v1.5.3
 
-RUN apt-get update &&\
-    apt-get install -y sudo git bash unattended-upgrades
-
 WORKDIR /code
 ENV PATH /usr/local/nvidia/bin:/usr/local/cuda/bin:/code/cmake-3.14.3-Linux-x86_64/bin:/opt/miniconda/bin:${PATH}
 ENV LD_LIBRARY_PATH /opt/miniconda/lib:$LD_LIBRARY_PATH
 
+RUN DEBIAN_FRONTEND=noninteractive \
```
Comment (triwahyuu) [inline dockerfiles/onnxruntime-tensorrt.dockerfile:19]: the `DEBIAN_FRONTEND` variable in here doesn't really affect anything really, as you can see from the build output.
I think the better usage in docker is to put it in `ARG` statement like the one in main dockerfile.

**Comment by alifahrri** [comment]
> it seems like you're not using paralel compilation when compiling the onnxruntime, that's why it takes way too long to complete.
> I think you could use the [`--parallel`](https://github.com/microsoft/onnxruntime/blob/f874260b9e0d26a603b64e15da5a454a4b67c229/tools/ci_build/build.py#L142) argument for it.
> reference from this https://github.com/microsoft/onnxruntime/blob/master/build.sh

Ok

File: dockerfiles/onnxruntime-tensorrt.dockerfile (hunk 4 above)
Comment (alifahrri) [reply dockerfiles/onnxruntime-tensorrt.dockerfile:30]: Ok

Comment (alifahrri) [reply dockerfiles/onnxruntime-tensorrt.dockerfile:30]: last time I tried using latest miniconda the installation for development will breaks, thats why it uses this version

Comment (alifahrri) [reply dockerfiles/onnxruntime-tensorrt.dockerfile:30]: but ok let me try again

Comment (triwahyuu) [reply dockerfiles/onnxruntime-tensorrt.dockerfile]: same actually with the main docker, but I use `--ignore-installed` to deal with it, just as a workaround really.

**Comment by triwahyuu** [comment]
well at least it way better than before. so could  you support for the hypopt as well for the development?

File: dockerfiles/onnxruntime-tensorrt.dockerfile [hunk 6]
```diff
@@ -7,39 +7,71 @@
...
+
+RUN curl -fsSL -v -o ~/miniconda.sh -O  https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh  && \
+    chmod +x ~/miniconda.sh && \
+    ~/miniconda.sh -b -p /opt/conda && \
+    rm ~/miniconda.sh && \
+    /opt/conda/bin/conda install -y python=${PYTHON_VERSION} && \
+    /opt/conda/bin/conda install -c pytorch -y cudatoolkit=10.2 && \
+    /opt/conda/bin/conda clean -ya
+RUN python -c "import sys; assert sys.version[:3] == '$PYTHON_VERSION', sys.version"
+RUN pip install numpy
```
Comment (triwahyuu) [inline dockerfiles/onnxruntime-tensorrt.dockerfile:39]: I think this one would be better to be included in the same command as conda installation together with pytorch. 
and also I think you forgot to include the conda path to env, like this `ENV PATH /opt/conda/bin:$PATH`

File: dockerfiles/onnxruntime-tensorrt.dockerfile [hunk 7]
```diff
@@ -7,39 +7,71 @@
...
+RUN cd src/runtime && pip install -U setuptools onnx==1.6.0 && pip install -e .
+
+# TODO: use CMD instead of RUN for proper testing
+RUN python -c "import vortex.runtime as vrt; assert(vrt.model_runtime_map['onnx']['tensorrt'].is_available())"
+
+FROM runtime AS development
+# install vortex-development
+COPY src/development src/development
+RUN pip install --upgrade pip
+RUN cd src/development && pip install --ignore-installed --timeout=10000 -e .
```
Comment (triwahyuu) [inline dockerfiles/onnxruntime-tensorrt.dockerfile:74]: wouldn't it be better to install it without `-e` flags? since the package inside the image most of the time is not used for development. also I think installing the optuna dependency doesn't really take that long

File: dockerfiles/onnxruntime-tensorrt.dockerfile (hunk 6 above)
Comment (alifahrri) [reply dockerfiles/onnxruntime-tensorrt.dockerfile:39]: I dont really agree to include this with conda since the python version assert is after conda, but yes I think that's subjective opinion

Comment (alifahrri) [reply dockerfiles/onnxruntime-tensorrt.dockerfile:39]: oh ya need to change miniconda to conda

Comment (triwahyuu) [reply dockerfiles/onnxruntime-tensorrt.dockerfile:39]: yepp I meant that one

**Comment by triwahyuu** [comment]
I think that's all for the the review, you can merge this

**Review by triwahyuu — APPROVED** [review]
